### ISSUE TITLE
[Windows] Loading SDL_TTF from a specific folder doesn't work

### ISSUE THREAD

**Issue by CodeMyst** [BindBC/bindbc-sdl#10]
I have a project where I'm using SDL2, SDL2_Image and SDL2_TTF. I downloaded the proper binaries from their site and placed them in the folder with the executable and everything works fine.

So I tried to move those sdl binaries to their own `lib` folder:
```
build/
    app.exe
    libs/
        - SDL2.dll
        - SDL2_image.dll
        - ...
```

And that doesn't work, every library loads except SDL2_TTF. I also tried loading only SDL2 and SDL2_TTF (without SDL2_Image) and that doesn't work either. So it looks like `loadSDLTTF` doesn't work when specifying the library path.

Here are also 2 example project, sdltest is with the lib folder and SDL_TTF doesn't load. sdltest2 is the one that does work (without lib folder):
* sdltest: https://drive.google.com/file/d/1DvkZG4hDi0k09nX0TcfdCIPy61eZrxng/view?usp=sharing
* sdltest2: https://drive.google.com/file/d/1PdJYyiAN_c205N_IQxF04ahUfFfb6Kit/view?usp=sharing

**Comment by mdparker** [comment]
Interesting. Error checking confirms that the system isn't finding it, but it also confirms that attempting to load the correct file path:

`lib\SDL2_ttf.dll: The specified module could not be found.`

So we know that the loader is passing the correct file name to `LoadLibrary`.  There's no obvious reason why it should be failing. Your guess is as good as mine.

**Comment by CodeMyst** [comment]
So someone said they had the same problem and that moving the binaries that SDL_TTF depends on to the same directory as the executable will fix it. So I moved only `libfreetype-6.dll` and `zlib1.dll` to the same directory as the executable and now it loads SDL_TTF correctly.
But SDL_Image has also a lot of dependencies and they work without having to move them to the same dir as the executable. 🤔

**Comment by mdparker** [comment]
Oh, I see. I tried it after moving the free type DLL to the exe directory but didn’t realize zlib is also a dependency.

So I believe the reason SDL_image gets away with it is that it loads it’s dependencies dynamically in the same way the BindBC loader does. I expect `LoadLibrary` will search based on the DLL’s module handle so it will look in the same directory the DLL is in. The SDL_ttf binary apparently is linked with the import libraries instead, so the dependent DLLs are loaded relative to the executable’s module handle and not SDL_ttf’s.

What you can do to work around it is to call `SetDllDirectory` before calling the load function. This will append a directory to the DLL search path. You’ll need to link with kernel32.dll.

https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setdlldirectorya

**Comment by CodeMyst** [comment]
Sorry for replying after such a long time, haven't worked on that part of the project in a while.

Doing `SetDllDirectory` does indeed work, but I also had to set `_WIN32_WINNT` to `0x0502` or higher. I did that by adding `"versions": "Windows7"` (dub.json) / `versions "Windows7"` (dub.sdl). That needs to be set to `Windows2003` or higher. https://github.com/dlang/druntime/blob/8fd52019826259dc92ab712f4b37a3f0ea4d8265/src/core/sys/windows/w32api.d#L26

I won't close the issue yet as a workaround is still needed.

**Comment by mdparker** [comment]
Revisiting this, I don't believe a workaround in bindbc is warranted. For two reasons.

1. This is only a problem on Windows. I could add a function to the base loader to append a directory to the search path, but it would only work on Windows (I don't know of any way to do it on Linux or Mac). I don't want any platform-specific functions in the API. 
2. It only occurs when dropping into a subdirectory DLLs that in turn load DLLs dynamically. That's a very specific (and I would venture, rare) use case.

I'll add some text to the README to warn of the issue.

**Comment by mdparker** [comment]
Actually, in the loader I could potentially check if the given path includes a subdirectory and, if so, append that to the path. But that means I have to account for relative vs. absolute paths, assume that the if it is relative then the user really wants me to load from the executable directory and not the working directory if it's different, an assumption that might not always be true. I'll leave this open longer while I consider it, but I still think it's rare enough that I should just leave it to the user.

**Comment by mdparker** [comment]
So I know it's been quite a while, but I've reconsidered my position on platform-specific functions. In this case, it actually makes sense. So I've added a function to the loader on Windows only: `setCustomLoaderSearchPath`. See the loader README for an example on how to use it:

https://github.com/BindBC/bindbc-loader#default-windows-search-path

Seems to do the trick in my local testing.

As I do maintenance updates to the other packages, I'll bump the loader dependency to 0.3.0 and update the SDL README to point folks to the new function.

Thanks for bringing it to my attention and sorry for the long delay!
